### PR TITLE
Start adding binaryuuid support.

### PR DESCRIPTION
**Pull request by dereuromark**
First stab at https://github.com/cakephp/phinx/issues/1732

I might need some help, as I am not too familiar with the lower levels here.

### CODE REVIEW

**Comment by dereuromark** [comment]
This now works for MySQL already:
```
CREATE TABLE `uuid16_test_rows` (
  `id` binary(16) NOT NULL,
  `name` varchar(100) COLLATE utf8_unicode_ci NOT NULL DEFAULT '',
  `foreign_key_without_index` binary(16) DEFAULT NULL,
  `foreign_key_with_index` binary(16) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

ALTER TABLE `uuid16_test_rows`
  ADD KEY `foreign_key_with_index` (`foreign_key_with_index`);
COMMIT;
```

**Comment by dereuromark** [comment]
Tests are green so far except SQLite

SQlite has some issues still
"General error: 1 near "`user_id`": syntax error"

uuid

> CREATE TABLE `tmp_ztable` (`id` UUID_TEXT NOT NULL, PRIMARY KEY (`id`), `user_id` INTEGER NOT NULL)

binaryuuid

> CREATE TABLE `tmp_ztable` (`id` BINARY_BLOB NOT NULL, PRIMARY KEY (`id`), `user_id` INTEGER NOT NULL)

**Comment by dereuromark** [comment]
The way SQLite creates those primary keys seems to be the issue
```
CREATE TABLE tmp_ztable
( id BINARY_BLOB NOT NULL PRIMARY KEY,
  user_id INTEGER NOT NULL
);
```
would be better

but that is a different topic and not directly related to the changes here.
So all good it seems.

I tested it also with CakePHP and migrations plugin, which now works correctly for all 3 db types: https://github.com/cakephp/migrations/issues/331

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 1]
```diff
@@ -443,6 +443,10 @@ public function getColumns($tableName)
                 $column->setPrecision($columnInfo['numeric_precision']);
             }
 
+            if ($column->getType() === static::PHINX_TYPE_BINARY && $column->getLimit() === 16) {
+                $column->setType(static::PHINX_TYPE_BINARYUUID);
+            }
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/PostgresAdapter.php:448]: Is it possible to enter this if statement?

Comment (dereuromark) [reply src/Phinx/Db/Adapter/PostgresAdapter.php:448]: Based on your findings it might not be necessary.

**Comment by MasterOdin** [comment]
For SQLServer, `binaryuuid` would probably map to the same type as `static::PHINX_TYPE_UUID`, namely `uniqueidentifier`, though like the postgresql adapter, you would not be able to do a comparison against `static::PHINX_TYPE_BINARYUUID` on saved columns in a table.

File: src/Phinx/Db/Adapter/PostgresAdapter.php (hunk 1 above)
Comment (markstory) [reply src/Phinx/Db/Adapter/PostgresAdapter.php:448]: Binary UUIDs in postgres would use the standard `UUID` type. Not a binary column.

Comment (dereuromark) [reply src/Phinx/Db/Adapter/PostgresAdapter.php:448]: That means we could safely remove this?

Comment (othercorey) [reply src/Phinx/Db/Adapter/PostgresAdapter.php:448]: I believe so. You should always use the uuid type with postgres.

Comment (dereuromark) [reply src/Phinx/Db/Adapter/PostgresAdapter.php]: Done

**Comment by dereuromark** [comment]
@MasterOdin For SqlServer it seems like UNIQUEIDENTIFIER could be the one to be used.
Do you want to look into this?
I think for the main DBs this support is sufficient for now.

File: src/Phinx/Db/Adapter/SQLiteAdapter.php [hunk 2]
```diff
@@ -56,6 +56,7 @@ class SQLiteAdapter extends PdoAdapter
         self::PHINX_TYPE_TEXT => 'text',
         self::PHINX_TYPE_TIME => 'time_text',
         self::PHINX_TYPE_UUID => 'uuid_text',
+        self::PHINX_TYPE_BINARYUUID => 'binary_blob',
```
Comment (dereuromark) [inline src/Phinx/Db/Adapter/SQLiteAdapter.php:59]: That seems to be the problem/wrong. See https://github.com/cakephp/migrations/issues/660

Comment (dereuromark) [reply src/Phinx/Db/Adapter/SQLiteAdapter.php:59]: uuid_blob is the correct one it seems

**Pull request merged**